src/crypto/equihash.cpp

```diff
@@ -72,15 +72,30 @@ eh_index UntruncateIndex(const eh_trunc t, const eh_index r, const unsigned int
 }
 
 template<size_t WIDTH>
-StepRow<WIDTH>::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i)
+StepRow<WIDTH>::StepRow(const eh_HashState& base_state, eh_index i,
+                        size_t hLen, size_t cBitLen, size_t cByteLen)
 {
     eh_HashState state;
     state = base_state;
     unsigned char array[sizeof(eh_index)];
     eh_index lei = htole32(i);
     memcpy(array, &lei, sizeof(eh_index));
     crypto_generichash_blake2b_update(&state, array, sizeof(eh_index));
-    crypto_generichash_blake2b_final(&state, hash, n/8);
+
+    if (8*cByteLen == cBitLen) {
+        // We are colliding an integer number of bytes; hLen == N/8
+        crypto_generichash_blake2b_final(&state, hash, hLen);
```

Comment:
Note that even when we change the expansion method, the same expression should work for the hash output length in both cases.